src/crimson/osd/heartbeat.cc

```diff
@@ -408,18 +509,189 @@ seastar::future<> Heartbeat::Peer::handle_reply(
   return seastar::now();
 }
 
-void Heartbeat::Peer::handle_reset(crimson::net::ConnectionRef conn)
+void Heartbeat::Peer::handle_reset(crimson::net::ConnectionRef conn, bool is_replace)
+{
+  if (con_front == conn) {
+    con_front = nullptr;
+    if (is_replace) {
+      assert(!front_ready);
+      assert(!session_started);
+      // set the racing connection, will be handled by handle_accept()
+      con_front = heartbeat.front_msgr->connect(
+          conn->get_peer_addr(), entity_name_t(CEPH_ENTITY_TYPE_OSD, peer));
+      has_racing = true;
+      logger().warn("Heartbeat::Peer::handle_reset(): "
+                    "con_front racing with osd.{}, updated by {}",
+                    peer, con_front);
+    } else {
+      if (front_ready) {
+        front_ready = false;
+      }
+      if (session_started) {
+        reset_session();
+      }
+      if (heartbeat.whoami == peer) {
+        logger().error("Heartbeat::Peer::handle_reset(): "
```

Comment:
does this happen in real world? how can we pick a peer whose id is identical to `whoami`? if this should never happen, can we have an `assert()` here instead?